File: bestia-client/src/Bnet/Message/Master/DeleteMasterCMSG.cs
using Bnet;
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Master
{
  /// <summary>
  /// Message for permanently deleting one of the account's masters.
  /// The result is reported back via OperationSuccess (MasterDeleted) / OperationError.
  /// </summary>
  public partial class DeleteMasterCMSG : ICMSG
  {
    [Export]
    public ulong MasterId { get; set; } = 0;

    /// <summary>
    /// The master name as the player typed it into the confirmation prompt. The server compares it against
    /// the real name and refuses the deletion on a mismatch, so never fill this in from the master data.
    /// </summary>
    [Export]
    public string ConfirmationName { get; set; } = "";

    public override Envelope ToEnvelope()
    {
      var deleteMaster = new global::Bnet.DeleteMasterCMSG
      {
        MasterId = MasterId,
        ConfirmationName = ConfirmationName
      };

      return new Envelope
      {
        DeleteMaster = deleteMaster
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/account/master/DeleteMasterCMSG.kt
package net.bestia.zone.account.master

import net.bestia.bnet.proto.DeleteMasterProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.util.AccountId

/**
 * Request to permanently delete one of the account's masters.
 *
 * Unlike most handlers this one cannot resolve its subject from the session
 * ([net.bestia.zone.ecs.core.session.ConnectionInfoService.getActiveEntityId]): deletion happens on the
 * character selection screen, before any master has been picked, so the id has to come from the client.
 * [MasterDeletionService] therefore checks ownership itself instead of relying on the session.
 */
data class DeleteMasterCMSG(
  override val playerId: AccountId,
  val masterId: Long,
  /** What the player typed into the confirmation prompt; must match the master's name. */
  val confirmationName: String
) : CMSG {
  companion object {
    fun fromBnet(accountId: AccountId, msg: DeleteMasterProto.DeleteMasterCMSG): DeleteMasterCMSG {
      return DeleteMasterCMSG(
        playerId = accountId,
        masterId = msg.masterId,
        confirmationName = msg.confirmationName
      )
    }
  }
}
